### PR TITLE
Initial ideas for new Encryption interface

### PR DESCRIPTION
**Pull request by milleruntime**
Looking for design feedback on a new encryption interface.  My thinking was to do something pluggable and seamless, similar to the way CompactionStrategy was done. 

If encryption is enabled call EncryptionStrategy.encryptStream() and EncryptionStrategy.decryptStream(), otherwise proceed normally. A DefaultEncryptionStrategy would be provided with basic encryption behavior.  Seems doable for BCFile, not so sure about the WALs.

Feedback PR only, not merging.

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 1]
```diff
@@ -45,78 +45,14 @@
   // Crypto-related properties
   @Experimental
   CRYPTO_PREFIX("crypto.", null, PropertyType.PREFIX,
-      "Properties in this category related to the configuration of both default and custom crypto"
-          + " modules."),
+      "Properties related to encryption."),
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Brevity! But, perhaps a little bit more would help. These are specifically "data-at-rest" or "file-based" encryption.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 2]
```diff
@@ -45,78 +45,14 @@
...
-  CRYPTO_OVERRIDE_KEY_STRATEGY_WITH_CONFIGURED_STRATEGY(
-      "crypto.override.key.strategy.with.configured.strategy", "false", PropertyType.BOOLEAN,
-      "The default behavior is to record the key encryption strategy with the"
-          + " encrypted file, and continue to use that strategy for the life of that"
-          + " file. Sometimes, you change your strategy and want to use the new"
-          + " strategy, not the old one. (Most commonly, this will be because you have"
-          + " moved key material from one spot to another.) If you want to override"
-          + " the recorded key strategy with the one in the configuration file, set"
-          + " this property to true."),
+  CRYPTO_RFILE_ENABLED("crypto.rfile.enabled", "false", PropertyType.BOOLEAN, "Enable encryption for R-Files."),
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Eventually, I think we want this to be per-table configuration.

File: core/src/main/java/org/apache/accumulo/core/security/crypto/EncryptionStrategy.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import org.apache.accumulo.core.conf.AccumuloConfiguration;
+import org.apache.accumulo.core.conf.Property;
+
+public interface EncryptionStrategy {
+
+  public void encryptStream(OutputStream outputStream);
+
+  public void decryptStream(InputStream inputStream);
+
+  public Map<String,String> properties = new HashMap<>();
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/security/crypto/EncryptionStrategy.java]: These properties shouldn't be public, and the getters/setters shouldn't have default methods.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java [hunk 4]
```diff
@@ -45,78 +45,14 @@
...
-      "Fully qualified class name of the class that implements the CryptoModule"
-          + " interface, to be used in setting up encryption at rest for the WAL and"
-          + " (future) other parts of the code."),
+  CRYPTO_STRATEGY("crypto.strategy", "org.apache.accumulo.core.security.crypto.DefaultEncryptionStrategy", PropertyType.CLASSNAME, "Encryption strategy"),
   @Experimental
-  CRYPTO_CIPHER_SUITE("crypto.cipher.suite", "NullCipher", PropertyType.STRING,
-      "Describes the cipher suite to use for rfile encryption. The value must"
-          + " be either NullCipher or in the form of algorithm/mode/padding, e.g."
-          + " AES/CBC/NoPadding"),
+  CRYPTO_WAL_ENABLED("crypto.wal.enabled", "false", PropertyType.BOOLEAN, "Enable encryption for Write Ahead Logs."),
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Could bake this (and the next property) into the strategy itself, and make the strategy aware of tables and wal contexts similar to how the VolumeChooser is aware of those.

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java (hunk 1 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: A little more for the description? Or you saying one of the properties should be "data-at-rest"?

Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: I'm saying 1) I like the shortening of the description, but 2) I think it should say something like "Properties related to on-disk file encryption" or "Properties related to encryption-at-rest" or whatever makes the most sense to clarify that it's separate from the RPC encryption properties.

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: Roger

File: core/src/main/java/org/apache/accumulo/core/conf/Property.java (hunk 2 above)
Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/conf/Property.java]: I agree.  I think for now I am going to make one "crypto.enabled" property.  So we can initialize the encryptionStrategy when its enabled.    I am thinking make the encryption per table as a follow on.

**Comment by milleruntime** [comment]
Closing this PR for now... will create another when I'm further along.

**Pull request closed without merging**